### ISSUE TITLE
cmd/go: {get|build|install} need to improve behaviors with vendor directory.

### ISSUE THREAD

**Issue by james-lawrence** [golang/go#44841]
### What version of Go are you using (`go version`)?

<pre>
go version
go version go1.16 linux/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

this is more of an experience report vs a specific bug/proposal around things I've found frustrating with go tooling since moving to modules for the last year.

### vendor setups are treated as second class citizens.
this shows itself in a number of ways.
1. in a repository that contains vendor directory dependencies `go get` doesn't automatically install to the vendor directory. requiring an additional command.
2. accesses the network to fetch dependencies when they're in the vendor directory.
3. numerous claims by go developers that go modules will allow for removing vendor support. this is not true and will never be true. in theory yes, you can build identical programs from go.mod and related infrastructure. however it ignores things like: reliance of 3rd party infrastructure or having to spinning up your own infrastructure to do so. vendoring is essentially that infrastructure in a single folder. it use to be far easier to use, understand, maintain, audit. however it has always been treated as as second class since support was added to satisfy the community at large.

### go get updates dependencies without explicit user action.
I've run into this a number of types where I'm working run so go tooling commands and when I build/run into an error where my vendor is out of sync with go.mod only to find a bunch of modules have been updated. Unfortunately I don't know exact which command is causing this, I suspect go get or go mod tidy but its just a gut feeling based on what I remember doing each time i've run into this. what I do know is this has occurred to me multiple times now where dependencies I was not interacting with in any way have been updated.

possibly related:
- https://github.com/golang/go/issues/40053 which spawned this [thread](https://groups.google.com/u/0/g/golang-tools/c/BRCgqwWLwoY) which i'm just now seeing because github never notified me on the updates to the issue.

### go modules has caused an explosion in complexity in the go environment

it has almost doubled the number of environment variable flags. its responsible for [9 out of 20](https://golang.org/cmd/go/#hdr-Environment_variables) environment variables.
- GOPROXY
- GONOPROXY
- GOPRIVATE
- GONOSUMDB
- GOSUMDB
- GOVCS
- GOINSECURE
- GOMODCACHE

most of which are accessing the network and all of which could essentially be dropped from a developers consciousness with a functional vendor directory. in fact if a vendor directory is detected in the main module during get/install/build you could in fact treat it as the gomodcache unless upgrading or adding a new package (in which case you're interacting with the VCS to populate that cache).

**Comment by kevinburkemeter** [comment]
Basically the way I work around this is by providing an extra command that sets all of the flags appropriately, and then just asking everyone to run `make vendor`.

```make
vendor:
	envdir envs/localdev go mod vendor && envdir envs/localdev go mod tidy
```

And then envs/localdev contains

```
$ ll envs/localdev | grep GO
-rw-r--r--  1 kevin staff   3 2019-08-14 16:21 GO111MODULE
-rw-r--r--  1 kevin staff  33 2019-10-04 12:12 GOFLAGS
-rw-r--r--  1 kevin staff 209 2020-01-16 21:49 GONOSUMDB
-rw-r--r--  1 kevin staff  19 2020-01-16 21:49 GOPRIVATE
-rw-r--r--  1 kevin staff   4 2020-01-16 21:49 GOSUMDB
```

in particular `GOFLAGS` is `-mod=vendor -race`, `GO111MODULE` set to `on`, and the other ones disable the public checksum db and cache.

**Comment by kevinburkemeter** [comment]
I'm not saying this is a good solution, it took a while to get there, but having it has simplified working with the vendor directory pretty well.

**Comment by james-lawrence** [comment]
@kevinburkemeter appreciate the workaround, looking for the discussion on how we as a community can fix the tool so it can provide a great experience out of the box without twiddling bits and bobs for the basic use cases. there is also a fundamental tooling problem underlying all of these problems which i don't think has been addressed @rsc touches on it in that thread I linked.
moving to a offline/read only starting point would resolve a ton of this complexity.

**Comment by bcmills** [comment]
> 1. in a repository that contains vendor directory dependencies `go get` doesn't automatically install to the vendor directory.  requiring an additional command.

That was an intentional decision resulting from the discussion in #30240 (see https://github.com/golang/go/issues/30240#issuecomment-524983085 for a summary).

The final design as implemented was #33848.

**Comment by bcmills** [comment]
> go get updates dependencies without explicit user action.

That behavior was changed in Go 1.16 (see proposal #42466). The `go.mod` file is no longer updated implicitly.

However, note that `go get` and `go mod tidy` _do_ continue to update the `go.mod` file, as that is the only purpose of those commands: they _are_ “explicit user action”. (If you want to install a binary without changing your dependencies, use `go install` instead of `go get`.)

**Comment by bcmills** [comment]
> 2. accesses the network to fetch dependencies when they're in the vendor directory.

That should never be the case when `-mod=vendor` is set (implicitly or explicitly).

When vendoring is in use, the only times the `go` command should access the network are when you run an explicit `go get` command (which has the _primary purpose_ of accessing the network to update your dependencies) or an explicit `go mod` subcommand (which has the _primary purpose_ of either resolving or reporting information that is not present in the `vendor` directory).

If you find that is not the case, please file a separate issue with steps to reproduce it.

**Comment by bcmills** [comment]
> 3. numerous claims by go developers that go modules will allow for removing vendor support.

As far as I am aware, no one on the Go project is claiming that modules will allow for removing vendor support. (If we were planning to remove it, we would not have bothered to implement #33848!)

Per the Background section on #33848:
> Go users have built a variety of workflows around vendored dependencies for a variety of use-cases — including self-contained builds, language-agnostic code review, and ephemeral patches — and since the launch of module mode in Go 1.11 they have made it clear that those workflows remain important in module mode.

**Comment by bcmills** [comment]
> in fact if a vendor directory is detected in the main module during get/install/build you could in fact treat it as the gomodcache unless upgrading or adding a new package (in which case you're interacting with the VCS to populate that cache).

Sorry, but that is not feasible. The `vendor` directory contains individual packages, not complete modules. It also does not in general include the `go.mod` files needed to reconstruct the dependencies of the modules containing those packages.

However, note that with `-mod=vendor` today the `go` command _already does_ use the `vendor` directory to load packages and dependency information “unless upgrading or adding a new package”. Also note that `-mod=vendor` has been the default behavior when a `vendor` directory is present since Go 1.14. (See https://golang.org/doc/go1.14#vendor.)

**Comment by bcmills** [comment]
Regarding the environment variables:

> GOPROXY
> GOSUMDB

These variables do admittedly add complexity. However, `GOPROXY` improves the availability of (non-vendored) dependencies, such as for new projects using existing libraries, and `GOSUMDB` improves security and reproducibility by ensuring that everyone agrees on the exact contents of the versions they require. We believe that those benefits substantially outweigh the cost of the added complexity.

> GONOPROXY
> GOPRIVATE
> GONOSUMDB

`GOPRIVATE` sets the defaults for `GONOPROXY` and `GONOSUMDB`. The latter are finer-grained controls, but most users should only need to bother with `GOPRIVATE`, which allows users to trade off the security and availability benefits mentioned above against the benefit of being able to work with code that is not publicly available.

> GOMODCACHE

The default value for `GOMODCACHE` is derived from `GOPATH` — this is another one of those variables that provides a finer-grained control for something that most users will probably set via a coarse-grained 

> GOINSECURE

`GOINSECURE` replaces the `-insecure` flag, which is deprecated (and planned for removal in Go 1.17) because it is _too_ insecure (#37519). So this one isn't _adding_ complexity so much as _replacing_ existing complexity, and it comes with a substantial increase in security.

> GOVCS

`GOVCS` is mostly unnecessary for public dependencies, because the public module proxy currently allows all of the supported VCS tools (with sandboxing to mitigate the security risks of less-well-tested tools).

(`GOVCS` is a bit like `GOINSECURE`, in that it allows for finer-grained security policies when updating private dependencies — improving security by default, but still allowing an opt-out for private dependencies on specific trusted hosts.)

**Comment by bcmills** [comment]
> this is more of an experience report vs a specific bug/proposal

As far as I can tell, many of the points here were either addressed, or explicitly considered and rejected, in #33848.

In order to revisit the behaviors rejected for #33848, we would need to see new information (something not already considered in the design and decisions for that proposal, such as a failure mode or mode of use that we were not aware of at the time), and probably also a concrete proposal of what to do differently.

**Comment by gudvinr** [comment]
@bcmills I have a library and supplementary tool which uses this library. These are different modules with different git repositories. Both library and the tool developed in parallel and I need to make changes in lib very often.

So, to test this library you need to make a commit then push and run `go get` in tool module. This is for library that sits basically within same parent directory. Every time you do so, debugger drops breakpoints since new library version goes to different directory. Update process is long since `go get` need to check remote git origin for changes.

I can't say how much time it'd save if I could only make symlink into vendor directory just for this particular library and nothing else.

**Comment by bcmills** [comment]
@gudvinr, that problem in general is #27542, which seems like it is (or ought to be) orthogonal to vendoring.

**Comment by james-lawrence** [comment]
@bcmills i'll get around to responding hopefully next weekend. but i think the problem with refering to #33848 is i believe the go team it making fundamental mistakes w/ respect to go modules that make it harder to work with golang tooling and defeating some of the main benefits to the tooling. the last few releases I've run into more and more problems with the tooling than at anytime in the 4 years prior.

and the problems start with how packages are resolved (and particularly the lack of a library for dealing with them), and progress into how all these commands interact with modules.

basically my intuition is telling me that the go team is taking a remote first approach to the resolution vs a local filesystem first approach. which is the source of most of the ergonomics of problems with vendoring. (-mod=vendor is a hack and unintuitive solution) and with the replace directive.

**Comment by james-lawrence** [comment]
@bcmills sorry for the huge delay in response. I just don't have the time to hand hold this issue I wish i could but life.

i do have a repository that doesn't build cleanly anymore due to go go modules treats replace directives and vendoring dependencies. you can use as an [example](https://github.com/james-lawrence/bw). unfortunately due to upstream motivations and I believe security issues its unlikely that the changes needed to support my use of said library will ever allow me to pull in upstream. i'm likely going to hard fork the dependency at some point as a result of lack of forward movement on this issue and #40053.

**Comment by bcmills** [comment]
I am not asking you to recreate an issue that was previously closed for a design point that was already settled. I am asking to to create a new issue if you have new information about something that isn't working as designed, or an aspect or implication of the design that was not known at the time the design was settled.

**Comment by seankhliao** [comment]
Closing as there is no actionable information here.
